lib/internal/modules/esm/loader.js

```diff
@@ -158,11 +158,19 @@ function nextHookFactory(chain, meta, validate) {
       if (generatedHookIndex === 0) { meta.chainFinished = true; }
 
       ArrayPrototypePush(args, nextNextHook);
-      const output = await ReflectApply(hook, undefined, args);
+      const output = ReflectApply(hook, undefined, args);
 
-      if (output?.shortCircuit === true) { meta.shortCircuited = true; }
-      return output;
+      function checkShortCircuited(output) {
+        if (output?.shortCircuit === true) { meta.shortCircuited = true; }
+      }
+
+      if (output instanceof Promise) { // eslint-disable-line node-core/prefer-primordials
+        output?.then(checkShortCircuited);
```

Comment:
`await` supports more than just `Promise`s, I suggest to keep supporting any thenable.
```suggestion
      const then = output?.then;
      if (typeof then === 'function') {
        ReflectApply(then, output, [checkShortCircuited, /* TODO: handle error case */]);
```

Reply:
In that case why not just PromiseResolve it unconditionally?

Reply:
@aduh95 I originally had something very similar to that but then switched to this because `(async () => 'foo')()` → `Promise<'foo'>`, and this seemed more straightforward/grokkable.

Reply:
@JakobJingleheimer well `instanceof` is a not reliable:

```js
{
  Object.defineProperty(Promise, Symbol.hasInstance, {
    __proto__: null,
    value: () => true,
  });

  console.log({} instanceof Promise); // true
}
```

@ljharb the goal of this PR is to make `resolve` fully synchronous, `PromiseResolve` is not an option.

Reply:
Sorry, missed Jordan's comment.

@aduh95 do we already have a helper for detecting a promise? If no, should I create one since it's not straightforward due to user-land?

Reply:
It could be a red-herring and not promise-related at all (as misguided as that would be).

Reply:
Yes, but that's how async/await works, so imo that's how we should keep it as well (btw the above load hooks works as is on current `master`)

Reply:
>btw the above load hooks works as is on current `master`

Sorry, what do you mean?

Reply:
I mean that the current implementation does support any thenables:

```console
$ out/Release/node --loader='data:text/javascript,export function load(url, context, next) { return { then(resolve) { resolve(next(url, context)); }}}' test/es-module/test-esm-assert-strict.mjs 
(node:72046) ExperimentalWarning: Custom ESM Loaders is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
$ out/Release/node --loader='data:text/javascript,export function load(url, context, next) { return { then(_, reject) { reject(new Error("thenable is rejected")); }}}' test/es-module/test-esm-assert-strict.mjs 
(node:72095) ExperimentalWarning: Custom ESM Loaders is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
data:text/javascript,export function load(url, context, next) { return { then(_, reject) { reject(new Error("thenable is rejected")); }}}:1
export function load(url, context, next) { return { then(_, reject) { reject(new Error("thenable is rejected")); }}}
                                                                             ^

Error: thenable is rejected
    at Object.then (data:text/javascript,export function load(url, context, next) { return { then(_, reject) { reject(new Error("thenable is rejected")); }}}:1:78)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

Node.js v19.0.0-pre
```

We should probably add tests for this kind of thenables, as well as tests where errors are thrown from the `load` hooks to make sure we cover that.

Reply:
👍Added to the todo list in the PR

Reply:
The reason to force it is that a non-Promise thenable might not behave properly, which is why the language is very careful to assimilate thenables into proper Promises asap.

`Promise.resolve(x) === x` is a way to detect native promises synchronously.

Reply:
As I said, I don't think we should try to detect if the hook returns a native promise or not, I think we should support any thenable that's "compliant with async/await". Do you have a suggestion to do that with `PromiseResolve`? 
An example would be great, as I don't see how `PromiseResolve` would help us here (we'd want something that works e.g. when `then` is a getter that throws on the second use).

```js
export function load(url, context, next) {
  let thenAlreadyAccessed = false;
  return {
    get then() {
      if (thenAlreadyAccessed) throw new Error;
      thenAlreadyAccessed = true;
      return resolve => resolve(next(url, context));
    }
  };
}
```

Reply:
I don’t think what you want is possible. The only thing compliant with async/await is wrapping a thenable into a native Promise.

I think the choices are “only accept native promises”, or “allow thenables to misbehave”, and the JS spec has tried very hard to disallow the latter.

Reply:
> We should probably add tests for this kind of thenables, as well as tests where errors are thrown from the `load` hooks to make sure we cover that.

1. Why are we throwing when these are received?
2. Can this be a follow-on PR improvement?

Reply:
> 1. Why are we throwing when these are received?

We are not, and we should not. We would be if we switch to using `instanceof Promise` checks though.



> 2\. Can this be a follow-on PR improvement?

I've opened https://github.com/nodejs/node/pull/43374 to address the missing tests. I don't think we should land this PR if can't pass the tests I'm suggesting to add – unless it turns out to be too complicated to support that it's not worth it.

Reply:
> We are not, and we should not. We would be if we switch to using instanceof Promise checks though.

Sorry, I wasn't clear. I meant, why are we doing this check? Why not just check that the hook's return is the expected shape and stop there? So for a sync `resolve`, look for `url` and maybe `format`, and throw on any other properties (such as `then`)? As in, why are we looking for one form of an invalid return, rather than validating a valid return?

Reply:
I went the route of specifically detecting the promise because it was previously a promise, so if a user updates Node and doesn't know `resolved` changed, "your resolve is wrong" is not so helpful (whereas "resolve isn't a promise anymore" is much better).